src/drivers/net/bnxt/bnxt.c

```diff
@@ -2123,17 +2245,257 @@ static void bnxt_adv_nq_index ( struct bnxt *bp, u16 cnt )
+#define BNXT_REG_WINDOW_BASE	0x400
+#define BNXT_GRC_BASE_MASK	0xfff
+#define BNXT_GRC_OFFSET_MASK	0xffc
+
+u32 bnxt_er_reg_write ( struct bnxt *bp, u32 reg_addr, u32 reg_val)
+{
+	u32 reg_base = 0;
+
+	reg_base = reg_addr & ~BNXT_GRC_BASE_MASK;
+
+	writel ( reg_base, bp->bar0 + BNXT_REG_WINDOW_BASE + 8 );
+
+	writel ( reg_val, bp->bar0 + ( BNXT_FW_HEALTH_WIN_OFF +
+				( reg_addr & BNXT_GRC_OFFSET_MASK ) ) );
+
+	DBGP ("bnxt_er_reg_write: reg_addr = %x, reg_val = %x\n", reg_addr, reg_val);
+	return reg_val;
+}
+
+u32 bnxt_er_reg_read ( struct bnxt *bp, u32 reg_addr)
+{
+	u32 reg_val = 0;
+	u32 reg_base = 0;
+
+	reg_base = reg_addr & ~BNXT_GRC_BASE_MASK;
+
+	writel ( reg_base, bp->bar0 + BNXT_REG_WINDOW_BASE + 8 );
+
+	reg_val = readl ( bp->bar0 + ( BNXT_FW_HEALTH_WIN_OFF +
+				( reg_addr & BNXT_GRC_OFFSET_MASK ) ) );
+
+	DBGP ("bnxt_er_reg_read: reg_addr = %x, reg_val = %x\n", reg_addr, reg_val);
+	return reg_val;
+}
+
+u32 bnxt_er_get_reg_val ( struct bnxt *bp, u32 reg_addr, u32 reg_type, u32 mask )
+{
+	u32 reg_val = 0;
+
+	switch ( reg_type ) {
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_PCIE_CFG:
+			pci_read_config_dword ( bp->pdev, reg_addr & mask, &reg_val );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_GRC:
+			reg_val = bnxt_er_reg_read ( bp, reg_addr );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR0:
+			reg_val = readl ( bp->bar0 + ( reg_addr & mask ) );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR1:
+			reg_val = readl ( bp->bar1 + ( reg_addr & mask ) );
+			break;
+		default:
+			break;
+	}
+	DBGP ( "read_reg_val bp %p addr %x type %x : reg_val = %x\n", bp, reg_addr, reg_type, reg_val );
+	return reg_val;
+}
+
+void bnxt_rst_reg_val ( struct bnxt *bp, u32 reg_addr, u32 reg_val )
+{
+	u32 mask = ER_QCFG_RESET_REG_ADDR_MASK;
+	u32 reg_type = reg_addr & ER_QCFG_RESET_REG_ADDR_SPACE_MASK;
+
+	switch ( reg_type ) {
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_PCIE_CFG:
+			pci_write_config_dword ( bp->pdev, reg_addr & mask, reg_val );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_GRC:
+			bnxt_er_reg_write ( bp, reg_addr, reg_val );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR0:
+			writel ( reg_val, bp->bar0 + ( reg_addr & mask ) );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR1:
+			writel ( reg_val, bp->bar1 + ( reg_addr & mask ) );
+			break;
+		default:
+			break;
+	}
+}
+
+void bnxt_rst_er_registers ( struct bnxt *bp )
+{
+	u32 delay_time = 0;
+	u8 i;
+
+	for ( i = 0; i < bp->er.reg_array_cnt; i++ ) {
+		bnxt_rst_reg_val ( bp, bp->er.rst_reg[i], bp->er.rst_reg_val[i] );
+
+		delay_time = bp->er.delay_after_rst[i];
+		if ( delay_time ) {
+			udelay ( delay_time * 100000 );
+		}
+	}
+
+}
+
+void bnxt_er_task ( struct bnxt* bp, u8 hb_task )
+{
+	u32 present_hb_cnt;
+	unsigned short pci_command, new_command;
+	u8 i;
+
+	DBGP ( "%s\n", __func__ );
+	if ( bp->er.er_rst_on ) {
+		/* Reset already in progress */
+		udelay ( bp->er.normal_wait_period * 100000 );
```

Comment:
This function seems to get called by `bnxt_poll()`, which is called in a tight loop by iPXE.  Why does it need to delay for some multiple of 100ms?

Reply:
Ack

Reply:
I have added a timer at the start of the error recovery code, and will use this as a means to keep track of the necessary wait time instead of using the long delay time

Reply:
@mcb30 - I have changed the hard delay to use a timer mechanism instead